File: Sprint2/Monitore/monitor-multiclass.java
// 1. Der Monitor
class Konto {

    private int kontostand = 0;

    public synchronized void einzahlen(int betrag) {
        kontostand += betrag;
        System.out.println("Eingezahlt: " + betrag + " | Stand: " + kontostand);
        notifyAll();
    }

    public synchronized void abheben(int betrag) {
        while (kontostand < betrag) {
            try {
                wait();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
        kontostand -= betrag;
        System.out.println("Abgehoben: " + betrag + " | Stand: " + kontostand);
    }
}

// 2. Klasse: Der Erzeuger
class Arbeitgeber implements Runnable {

    private Konto k;

    public Arbeitgeber(Konto k) {
        this.k = k;
    }

    public void run() {
        while (true) {
            k.einzahlen(50); // zahlt alle 2s, 50 € ein
            try {
                Thread.sleep(2000); // pausiert 2s
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }
}

// 3. Klasse: Der Verbraucher
class Arbeitnehmer implements Runnable {

    private Konto k;

    public Arbeitnehmer(Konto k) {
        this.k = k;
    }

    public void run() {
        while (true) {
            /*
            versucht Geld abzuheben -> erst möglich wenn mindestens 100 € eingezahlt
            -> Thread wird schlafen gelegt bis er von einzahlen() geweckt wird
            und anschließend die Bedingung in abheben() erfüllt ist
            */
            k.abheben(100);
        }
    }
}

File: Sprint2/Monitore/monitor-multiclass.py
import threading
import time


# 1. Der Monitor
class Konto:
    def __init__(self):
        self.kontostand = 0
        self.cond = threading.Condition()

    def einzahlen(self, betrag):
        with self.cond:
            self.kontostand += betrag
            print(f"Eingezahlt: {betrag} | Stand: {self.kontostand}")
            self.cond.notify_all()

    def abheben(self, betrag):
        with self.cond:
            while self.kontostand < betrag:
                self.cond.wait()
            self.kontostand -= betrag
            print(f"Abgehoben: {betrag} | Stand: {self.kontostand}")


# 2. Der Erzeuger
class Arbeitgeber(threading.Thread):
    def __init__(self, konto):
        super().__init__()
        self.konto = konto

    def run(self):
        while True:
            # zahlt alle 2s 50 € ein
            self.konto.einzahlen(50)
            time.sleep(2)


# 3. Der Verbraucher
class Arbeitnehmer(threading.Thread):
    def __init__(self, konto):
        super().__init__()
        self.konto = konto

    def run(self):
        while True:
            # versucht Geld abzuheben -> erst möglich wenn mindestens 100 € eingezahlt
            # -> Thread wird schlafen gelegt bis er von einzahlen() geweckt wird
            # und anschließend die Bedingung in abheben() erfüllt ist
            self.konto.abheben(100)
